agent/lib/kontena/workers/container_health_check_worker.rb

```diff
@@ -107,7 +107,10 @@ def check_tcp_status(ip, port, timeout)
     end
 
     def restart_container
-      Kontena::ServicePods::Restarter.new(@container.service_id, @container.instance_number).perform
+      Celluloid::Notifications.publish('service_pod:restart', {
```

Comment:
Sending this as a celluloid notification isn't necessarily better than just restarting the container: if the notification gets queued up over a `PodWorker` apply that re-creates/starts the container, then the container can end up getting restarted unnecessarily.

Maybe the `service_pod:restart` notification needs a `at: Time.now`, and it can get ignored if the container was started after the restart notification was sent?

Reply:
Fixed this to also include the `@container.id` and `@container.started_at` in the notification... the `ServicePodWorker#restart` will ignore the request if the container was already re-created or re-started.

Reply:
This being async means that the `ContainerHealthCheckWorker` will now continue doing healthchecks and notifying restarts for a container even after it determines it to be unhealthy. That's not really a problem, because the `ServicePodWorker` will explicitly ignore those stale restart notifications:

```
D, [2018-02-09T09:19:02.549742 #1] DEBUG -- Kontena::Workers::ServicePodWorker: mark healthcheck-restart/redis-3 for restart at 2018-02-09 09:19:02 +0000
D, [2018-02-09T09:19:02.551441 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of healthcheck-restart.redis-3: running
D, [2018-02-09T09:19:03.617821 #1] DEBUG -- Kontena::Workers::ContainerHealthCheckWorker: checking health for container: /healthcheck-restart.redis-3 using tcp ip and port: 10.81.128.60:6378
I, [2018-02-09T09:19:03.657576 #1]  INFO -- Kontena::Workers::ServicePodWorker: restarting healthcheck-restart.redis-3
I, [2018-02-09T09:19:03.793155 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: About to restart container healthcheck-restart.redis-3 as it's reported to be unhealthy
I, [2018-02-09T09:19:04.395778 #1]  INFO -- Kontena::ServicePods::Restarter: restarting service: /healthcheck-restart/redis-3
D, [2018-02-09T09:19:04.650668 #1] DEBUG -- Kontena::Workers::ContainerHealthCheckWorker: checking health for container: /healthcheck-restart.redis-3 using tcp ip and port: 10.81.128.60:6378
I, [2018-02-09T09:19:04.757311 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: About to restart container healthcheck-restart.redis-3 as it's reported to be unhealthy
D, [2018-02-09T09:19:05.651397 #1] DEBUG -- Kontena::Workers::ContainerHealthCheckWorker: checking health for container: /healthcheck-restart.redis-3 using tcp ip and port: 10.81.128.60:6378
I, [2018-02-09T09:19:05.651910 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: About to restart container healthcheck-restart.redis-3 as it's reported to be unhealthy
D, [2018-02-09T09:19:06.699934 #1] DEBUG -- Kontena::Workers::ContainerHealthCheckWorker: checking health for container: /healthcheck-restart.redis-3 using tcp ip and port: 10.81.128.60:6378
I, [2018-02-09T09:19:06.914658 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: About to restart container healthcheck-restart.redis-3 as it's reported to be unhealthy
D, [2018-02-09T09:19:07.730113 #1] DEBUG -- Kontena::Workers::ContainerHealthCheckWorker: checking health for container: /healthcheck-restart.redis-3 using tcp ip and port: 10.81.128.60:6378
I, [2018-02-09T09:19:08.294620 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: About to restart container healthcheck-restart.redis-3 as it's reported to be unhealthy
I, [2018-02-09T09:19:08.294918 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: stopped to check status from /healthcheck-restart.redis-3
D, [2018-02-09T09:19:09.093401 #1] DEBUG -- Kontena::Workers::ServicePodWorker: skip healthcheck-restart/redis-3 state sync at 2018-02-09 09:13:39 UTC: unchanged
D, [2018-02-09T09:19:09.457247 #1] DEBUG -- Kontena::Workers::ServicePodWorker: stale healthcheck-restart/redis-3 restart for container started_at=2018-02-09T09:18:59+00:00
D, [2018-02-09T09:19:09.477892 #1] DEBUG -- Kontena::Workers::ServicePodWorker: healthcheck-restart/redis-3 stale container event: kill at 2018-02-09T09:19:04.472038030Z < 2018-02-09T09:19:07.441366524+00:00
D, [2018-02-09T09:19:09.482319 #1] DEBUG -- Kontena::Workers::ServicePodWorker: stale healthcheck-restart/redis-3 restart for container started_at=2018-02-09T09:18:59+00:00
I, [2018-02-09T09:19:09.344934 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: starting to watch health of container /healthcheck-restart.redis-3
D, [2018-02-09T09:19:09.482718 #1] DEBUG -- Kontena::Workers::ServicePodWorker: healthcheck-restart/redis-3 stale container event: die at 2018-02-09T09:19:04.893637418Z < 2018-02-09T09:19:07.441366524+00:00
D, [2018-02-09T09:19:09.654705 #1] DEBUG -- Kontena::Workers::ServicePodWorker: stale healthcheck-restart/redis-3 restart for container started_at=2018-02-09T09:18:59+00:00
D, [2018-02-09T09:19:09.655092 #1] DEBUG -- Kontena::Workers::ServicePodWorker: healthcheck-restart/redis-3 stale container event: stop at 2018-02-09T09:19:05.933237314Z < 2018-02-09T09:19:07.441366524+00:00
```